File: src/main/java/com/micahelias/opengl/VertexBufferLayout.java
package com.micahelias.opengl;

import java.util.ArrayList;
import static org.lwjgl.opengl.GL11.*;

class VertexBufferElement {
  int type;
  int count;
  boolean normalized;

  public VertexBufferElement(int type, int count, boolean normalized) {
    this.type = type;
    this.count = count;
    this.normalized = normalized;
  }

  public int getTypeSize() {
    switch(type){
      case GL_FLOAT:
        return Float.BYTES;
      case GL_UNSIGNED_INT:
        return Integer.BYTES;
      case GL_UNSIGNED_BYTE:
        return Character.BYTES;
      default:
      return 0;
    }
  }

}

public class VertexBufferLayout {

  ArrayList<VertexBufferElement> m_Elements = new ArrayList<VertexBufferElement>();
  int m_Stride = 0;

  public void pushf(int count) {
    m_Elements.add(new VertexBufferElement(GL_FLOAT, count, false));
    m_Stride += Float.BYTES * count;
  }

  public void pushi(int count) {
    m_Elements.add(new VertexBufferElement(GL_UNSIGNED_INT, count, false));
    m_Stride += Integer.BYTES * count;
  }


  public void pushc(int count) {
    m_Elements.add(new VertexBufferElement(GL_UNSIGNED_BYTE, count, true));
    m_Stride += Character.BYTES * count;
  }

  public ArrayList<VertexBufferElement> getElements() {
    return m_Elements;
  }

  public int getStride() {
    return m_Stride;
  }



}

File: mango/opengl/vertex_buffer_layout.go
package opengl

import "github.com/go-gl/gl/v3.3-core/gl"

type VertexBufferLayout struct {
	m_Elements []VertexBufferElement
	m_Stride   int
}

func NewVertexBufferLayout() *VertexBufferLayout {
	return &VertexBufferLayout{}
}

func (vbl *VertexBufferLayout) Pushf(count int) {
	vbl.m_Elements = append(vbl.m_Elements, VertexBufferElement{gl.FLOAT, count, false})
	vbl.m_Stride += 4 * count
}

func (vbl *VertexBufferLayout) Pushi(count int) {
	vbl.m_Elements = append(vbl.m_Elements, VertexBufferElement{gl.UNSIGNED_INT, count, false})
	vbl.m_Stride += 4 * count
}

func (vbl *VertexBufferLayout) Pushc(count int) {
	vbl.m_Elements = append(vbl.m_Elements, VertexBufferElement{gl.UNSIGNED_BYTE, count, true})
	vbl.m_Stride += 1 * count
}

func (vbl *VertexBufferLayout) GetElements() []VertexBufferElement {
	return vbl.m_Elements
}

func (vbl *VertexBufferLayout) GetStride() int {
	return vbl.m_Stride
}
